---
title: "Vitamin D - COVID-19 MR (redo)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{VitaminD_COVID_MR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
params: 
  outcome_gwas: COVID-A2
  exposure_name: Vitamin D
  prefix: VD_A2 
  pop: !r c("CEU", "TSI", "GBR", "IBS")
  LDLink_token: !r tryCatch(glue::trim(readChar(here::here("LD_LINK_TOKEN"), 20)), warning=function(cond) { message("The file LD_LINK_TOKEN should contain an access token to the NCI API. if you don't have a token go here:\n https://ldlink.nci.nih.gov/?tab=apiaccess"); return(""); })
  near_half_threshold: 0.08
  exposure_gwas_file: ./1-s2.0-S0002929720300173-mmc2.xlsx
  outcome_file: ./COVID19_HGI_A2_ALL_leave_23andme_20201020.b37.txt.gz
  exposure_pvalue_threshold: !r as.numeric("5e-5") 
  EAF_rare_threshold: 0.01
  prune_r2_threshold: 0.05
  high_LD_threshold: 0.8
  far_SNP_proxy: 500000
  minP: !r as.numeric(1e-300)
  skip_LD_api: !r TRUE
  results_dir: ../derived_data
  use_cache: TRUE
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>")
  
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = params$use_cache)
knitr::opts_chunk$set(warning = TRUE)
#knitr::opts_chunk$set(error = TRUE)
knitr::opts_chunk$set(comment = "")

```

```{r setup}

library(MRutils)
```


```{r show parameters}
for (i in names(params)) { 
  if (i != "LDLink_token" || params[i]=="") 
    print(glue::glue("{i} = {params[i]}")) 
  else
    print(glue::glue("{i} = [******]"))
}

```


## Extract data from Exposure GWAS

```{r extract excelsheet data, verbose=TRUE}
# Since the input data is given to us as an excel spreadsheet, we have to extract the information from it
# "manually". In order to avoid copy-paste errors, we extract the information using the readxl library. 

  exposure_file=params$exposure_gwas_file
  if(!file.exists(exposure_file)){
    stop(glue::glue("cannot find file {exposure_file}, current dir is {getwd()}"))
  }
  data = readxl::read_excel(exposure_file,
                  sheet = "Table S2", 
                  skip = 4, 
                  .name_repair = "minimal",col_types = c("text")) %>% 
  subset(select = c(1:5, 13:17))

first_blank = min(which(is.na(data$RSID)))
# use the first blank line as an indication of the end of the data, since the dictionary comes after
if (length(first_blank) == 1) {
  data = data[1:min(first_blank - 1, nrow(data)),]
}

text=textConnection(object = NULL, open = "w")
write.table(data,text,quote = F,sep = "\t",row.names = F)
data=data.table::fread(text=textConnectionValue(text))

names(data)[8] = "beta"

names(data)[1] = "rsid"
data=data %>% 
  dplyr::mutate_at(.vars = c("EAF", "beta", "SE", "P"), as.numeric) %>% 
  dplyr::mutate(POS = BP, CHR = as.character(gsub("^chr", "", CHR)), BP = NULL) %>%
  dplyr::mutate(P = pmax(P, params$minP))
  

```
Read `r nrow(data)` entries from `r exposure_file`


Making sure that the exposure data validates as a gwas...`r if(assert_gwas(data, on_error='tell')) "yes" else"no"`
```{r Validate exposure data, results='hide'}
assert_gwas(data)
```

```{r Filter significant SNPs}
significant_snps <- filter_and_write_exposure_data(
  data,
  location_prefix = system.file("derived_data", params$prefix, package = "MRutils"),
  params$exposure_pvalue_threshold,
  params$EAF_rare_threshold
  )
  
```
Selecting significant snps and those that are not very rare (or common)...
found `r nrow(significant_snps)` significant and common SNPs.

##TODO: Add a quality control step that compares EAFs to dbSNP MAFs. and possibly filters deviating SNPs (by percent) 

<!-- json_datum=jsonlite::read_json("https://rest.ensembl.org/variation/human/rs10832310?pops=1;content-type=application/json") -->
<!-- json_datum$populations%>%Filter(f = function(x)grepl("1000GENOMES", x$population))%>%Filter(f = function(x)grepl("EUR", x$population)) -->

## Outcome  

### Extract relevant SNPs
```{r extract Snps}

map_outcome_data <- function(x) {
  
  dplyr::mutate(x,
         CHR = as.character(gsub("^chr", "", X.CHR)),
         beta = all_inv_var_meta_beta, 
         P = all_inv_var_meta_p, 
         EA = ALT, 
         NEA = REF, 
         EAF = all_meta_AF, 
         SE = all_inv_var_meta_sebeta) %>% 
      subset(select = required_headers)
}
 
 show(getwd)
  outcome_data <- extract_snps_from_bgzip(outcome = params$outcome_file, 
                                          snps = significant_snps, 
                                          comment_char = "", 
                                          mapping_function = map_outcome_data,
                                          validate = FALSE)

  rsIdsFromDBSNP = glue::glue("{params$results_dir}/{params$prefix}_rsIdsFromdbSnp.txt")
  
```
Making sure that the outcome data validates as a gwas...`r if(assert_gwas(outcome_data, on_error='tell')) "yes" else "no" `

`r glue::glue("{nrow(outcome_data)} exposure SNPs were found in the outcome file {params$outcome_file}")`

```{r assert, results='hide'}
assert_gwas(outcome_data)
```
### Find rsids for un-named snps
`r glue::glue("There are {nrow(outcome_data %>% subset(is.na(rsid)))} SNPs in the outcome data that are missing their rsid: ")`

```{r }
outcome_data %>% subset(is.na(rsid), select = c(CHR, POS, EA, NEA))
```

```{r find_unnamed_snps, cache=params$use_cache, results='hide'}
## this can take time and hits the API multiple times....
outcome_with_rsids <- fill_gwas_unknown_rsids(outcome_data)
  
```
`r glue::glue("After merging the results, there are {nrow(outcome_with_rsids%>%subset(is.na(rsid)))} variants lacking rsid, and {nrow(outcome_with_rsids)} variants overall.")`

```{r merge exposure and outcome}
merged_exp_and_out <- merge(significant_snps, suffixes = c(".exp", ".out"),
                          outcome_with_rsids, 
                          by = c("rsid", "CHR", "POS"), 
                          all.x = TRUE,
                          )

```
`r glue::glue("After merging Exposure and Outcome, there are {nrow(merged_exp_and_out%>%subset(is.na(beta.out)))} variants lacking 'beta.out', and {nrow(merged_exp_and_out)} variants overall.")`

```{r get proxies, cache=params$use_cache}

##takes time (hits the LDLink API)
# proxies_raw <- MRutils::get_proxies_given_candidates(variant_query = subset(merged_exp_and_out, is.na(beta.out)),
#                                                      variant_candidates_file = params$outcome_file,
#                                                      token = params$LDLink_token,
#                                                      comment_char = "",
#                                                      mapping_function = map_outcome_data ,
#                                                      r2_threshold = params$high_LD_threshold,
#                                                      candidate_distance = params$far_SNP_proxy,
#                                                      population = params$pop )

proxies_raw <- MRutils::get_proxies(rsids = subset(merged_exp_and_out, is.na(beta.out))$rsid,
                                                   token = params$LDLink_token,
                                                   results_dir = "../derived_data",
                                                   r2_threshold = params$high_LD_threshold,
                                                   population = params$pop )

```
`r glue::glue("Found {nrow(proxies_raw)} proxies for {length( unique(proxies_raw$query_rsid))} variants, out of {length(which(is.na(merged_exp_and_out$EAF.out)))} requested")`

Subset to proxies that are in the Outcome, since we need to find out which proxies are present there.
```{r use proxies}
proxies_in_outcome <- extract_snps_from_bgzip(params$outcome_file, snps = proxies_raw, mapping_function = map_outcome_data)

rsIdsFromDBSNP_proxies = glue::glue("{params$results_dir}/{params$prefix}_proxy_rsIdsFromdbSnp.txt")
```
`r glue::glue("need to look for rsid for {nrow(subset(proxies_in_outcome, is.na(rsid)))} SNPs")`


```{r find proxies_missing_rsids, cache=params$use_cache, results='hide'}
proxies_in_outcome_with_missing_ids <- get_unknown_rsids_from_locus(proxies_in_outcome)

write.table(x = proxies_in_outcome_with_missing_ids, file = rsIdsFromDBSNP_proxies, quote = F, sep = '\t', row.names = F)
```

### Merge proxies into Outcome
```{r merge with rest of outcome}
proxies_in_outcome_with_missing_ids <- read.table(file = rsIdsFromDBSNP_proxies,  sep = '\t', header = TRUE)

proxies_in_outcome_with_rsids <- merge_rsids_into_gwas(proxies_in_outcome, proxies_in_outcome_with_missing_ids)

proxies <- subset(proxies_raw, rsid %in% proxies_in_outcome_with_rsids$rsid)

```
Found rsids for `r nrow(proxies_in_outcome_with_missing_ids)` SNPs.

There are `r nrow(proxies)` proxies (of `r nrow(proxies_raw)`) in the outcome.

Merge proxies with exposure...
```{r merge proxies and exposure}
exposure_and_proxies <- merge(proxies, significant_snps, by.x = c("query_rsid","CHR"), by.y = c("rsid","CHR"), suffixes = c(".proxy",".exp"))
```
... `r nrow(exposure_and_proxies)` proxies were successfully merged with the exposure.

Fix alleles
```{r fix-proxy-alleles}
fixed_proxies_with_exposure <- fix_proxy_alleles(exposure_and_proxies)
```

### TODO: consider choosing the best snp only AFTER merging with the outcome so that the max AF can be used to filter out the palindromic snps

Remove palindromic snps and then the smallest P
```{r, remove palindromic SNPs}
fixed_proxies_with_exposure <- choose_best_proxies(fixed_proxies_with_exposure, params$near_half_threshold)
```

After removing ambiguous snps and choosing the most linked SNP for every query, `r nrow(fixed_proxies_with_exposure)` proxies remain.

```{r Check alleles, eval=FALSE}
with(fixed_proxies_with_exposure, Alleles == glue::glue("({EA}/{NEA})") | Alleles == glue::glue("({NEA}/{EA})"))
subset(fixed_proxies_with_exposure, select = c(Alleles, Correlated_Alleles, rsid, EA, NEA))
```


```{r merge proxy rows with outcome and combine with rest}
merged_exp_and_out_with_query <- dplyr::mutate(merged_exp_and_out, query_rsid = NA)
merged_with_proxies <- merge(proxies_in_outcome_with_rsids,
                             fixed_proxies_with_exposure,
                             suffixes = c(".out",".exp"),
                             by.x = c("rsid", "CHR"),
                             by.y = c("rsid", "CHR")) %>%
  dplyr::select(names(merged_exp_and_out_with_query))

merged_and_combined <- rbind(merged_with_proxies, 
                             merged_exp_and_out_with_query, 
                             stringsAsFactors = FALSE) %>%
  subset(!is.na(EAF.out))
```

Total number of SNPs for use in 2-Sample MR (prior to pruning): `r nrow(merged_and_combined)`

Now we need to make sure that we do not have SNPs in high LD next to each other.
Pruning using populations: [`r glue::glue_collapse(glue::glue("{params$pop}"),sep=", ")`]
```{r prune_snps, cache = TRUE, results='hide'}

pruning_results <- remove_linked_snps(merged_and_combined, pop = params$pop, token = params$LDLink_token, params$prune_r2_threshold)

```

```{r}
results_for_MR_file <- glue::glue("{params$results_dir}/{params$prefix}_SNPsFor2SMR.txt")
write.table(pruning_results$pruned, file = results_for_MR_file, sep = '\t', quote = F, row.names = F)
```
Removed `r length(pruning_results$removed_rsids)` snps due to high LD: `r glue::glue_collapse(glue::glue("{pruning_results$removed_rsids} "), sep="\n")`

Total number of SNPs for use in 2-Sample MR (pruned): `r nrow(pruning_results$pruned)`

```{r get MR results}
results <- get_2smr_results(results_for_MR_file)
```

```{r show results}
if(require(ggplot2)){
    
  for (p in results) {
    show(p)
  }
}
save(results, file = glue::glue("{params$results_dir}/{params$prefix}_2SMR_results.RData"))

```

```{r compare with Guillaumes results, eval=FALSE} 
guillaume=rbind(
  data.frame(rsid=c("rs6698680", "rs3750296", "rs7519574", "rs56044892", "rs2934744", "rs7528419", "rs3768013", "rs115045402", "rs144613541", "rs61816761", "rs11264360", "rs867772"), CHR="1"),
  data.frame(rsid=c("rs12997242", "rs11127048", "rs6724965", "rs7569755", "rs1047891", "rs2011425"), CHR="2"),
  data.frame(rsid=c("rs1972994", "rs6438900", "rs6773343"), CHR="3"),
  data.frame(rsid=c("rs78649910", "rs7699711", "rs11723621", "rs58073039", "rs28364331", "rs1229984"), CHR="4"),
  data.frame(rsid="rs7718395", CHR="5"),
  data.frame(rsid="rs3822868", CHR="6"),
  data.frame(rsid=c("rs111529171", "rs1011468", "rs1858889"), CHR="7"),
  data.frame(rsid=c("rs804280", "rs34726834", "rs7828742"), CHR="8"),
  data.frame(rsid=c("rs10818769", "rs532436 (proxy: rs635634)"), CHR="9"),
  data.frame(rsid=c("rs10887718"), CHR="10"),
  data.frame(rsid=c("rs10832218", "rs577185477", "rs10832289", "rs188480917 (proxy: rs576128895) ", "rs117576073", "rs523583", "rs12803256", "rs200454003 (proxy: rs4245442)", "rs10793129", "rs1149605", "rs964184", "rs2847500", "rs12317268"), CHR="11"),
  data.frame(rsid=c("rs9668081", "rs10859995"), CHR="12"),
  data.frame(rsid=c("rs8018720"), CHR="14"),
  data.frame(rsid=c("rs261291", "rs1800588","rs17765311", "rs62007299"), CHR="15"),
  data.frame(rsid=c("rs8063706", "rs77924615", "rs1800775"), CHR="16"),
  data.frame(rsid=c("rs2909218"), CHR="17"),
  data.frame(rsid=c("rs8091117", "rs2037511"), CHR="18"),
  data.frame(rsid=c("rs57631352", "rs73015021", "rs10500209", "rs58542926", "rs3814995", "rs1065853", "rs157595", "rs112285002 (proxy: rs296386)", "rs10426", "rs8103262"), CHR="19"),
  data.frame(rsid=c("rs6123359", "rs6127099", "rs2585442", "rs2762942"), CHR="20"),
  data.frame(rsid=c("rs2229742"), CHR="21"),
  data.frame(rsid=c("rs2074735", "rs960596"), CHR="22"))%>%mutate(rsid=gsub(rsid, pattern = " .*",replacement = ""))

###investigate differences:

subset(pruning_results$pruned$rsid, rsid %notin% guillaume$rsid)
subset(guillaume, rsid %notin% pruning_results$pruned$rsid)

rsids_differences=rbind(data.frame(subset(guillaume,rsid %in% setdiff(guillaume$rsid, pruning_results$pruned$rsid)),set="Y"),
                        data.frame(subset(pruning_results$pruned, rsid %in% setdiff(pruning_results$pruned$rsid, guillaume$rsid), select=c("rsid","CHR")), set="G"))

subset(proxies, rsid %in% rsids_differences$rsid)


ld_pairs_raw2 <- get_LD_pairs(rsids_differences, population = params$pop, token = params$LDLink_token)
ld_pairs_raw2%<>%subset(value>0.5 & variable!=RS_number)

merge(merge(ld_pairs_raw2,rsids_differences, by.x=c("RS_number","CHR"), by.y=c("rsid","CHR")), rsids_differences, by.x=c("variable","CHR"), by.y=c("rsid","CHR"))


```


```{r eval=FALSE}
# missing="rs2585442" ## Guillaume agrees that it should be removed since it has r^2 0.52 with rs6127099
# missing="rs11723621" # instead of this SNP, we could keep rs705117 & rs560384646
# missing="rs10832289" # I chose a different LD linked snp rs10832310 but it's palindromic :-(
# missing="rs12803256" # I chose a different LD linked snp rs12808183
# missing="rs200454003" # This snp's position 11:71228990 isn't in the A2 outcome gwas...not sure what's going on 
# missing="rs112285002" # This snp's position 19:48374320 isn't in the A2 outcome gwas but it is in the C2 outcome.
# missing="rs2585442" # I chose a different LD linked snp rs6127099

subset(significant_snps, rsid==missing)
subset(outcome_data, rsid==missing)
subset(outcome_with_rsids, rsid==missing)
subset(proxies, rsid==missing)
subset(proxies_in_outcome_with_missing_ids, rsid==missing)
subset(proxies_in_outcome_with_rsids, rsid==missing)
subset(proxies, rsid==missing)
subset(merged_exp_and_out, rsid==missing)
subset(proxies, rsid==missing)
subset(exposure_and_proxies, rsid==missing)
subset(proxies_with_exposure, rsid==missing)
subset(fixed_proxies_with_exposure, rsid==missing)
subset(merged_with_proxies, rsid==missing)
subset(merged_and_combined, rsid==missing)
subset(pruning_results$pruned, rsid==missing)
missing %in% removed_rsids
subset(ld_pairs_raw,RS_number==missing | value==missing )
subset(ld_pairs_raw,(RS_number==missing | value==missing ) & value >0.05)
subset(ld_pairs,(RS_number==missing | value==missing ) & value >0.05 & RS_number!=variable)
subset(ld_pairs,(RS_number==missing | value==missing ) & value >0.05 & RS_number!=variable)$variable %in% pruning_results$pruned$rsid
subset(pruning_results$pruned,rsid %in% subset(ld_pairs,(RS_number==missing | value==missing ) & value >0.05 & RS_number!=variable)$variable)
subset(merged_and_combined,rsid %in% subset(ld_pairs,(RS_number==missing | value==missing ) & value >0.05 )$variable)


```